configmap/dynamic.go

```diff
@@ -0,0 +1,250 @@
+		d.refreshCancel <- struct{}{}
+	}
+}
+
+func (d *dynamicConfigInternal) getCurrent() *ConfigState {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	return d.current
+}
+
+func (d *dynamicConfigInternal) loadAndTryFixConfig() (map[string]string, error) {
+	configMapsClient := d.client.CoreV1().ConfigMaps("default")
+	controllerConfig, err := configMapsClient.Get(context.TODO(), d.configMapName, metaV1.GetOptions{})
+	if err != nil {
+		if d.options.Template == nil {
+			return nil, err
+		}
+		if serr, ok := err.(*errors.StatusError); !ok || serr.ErrStatus.Reason != metaV1.StatusReasonNotFound {
+			return nil, err
+		}
+		configMap := corev1.ConfigMap{
+			ObjectMeta: metaV1.ObjectMeta{
+				Namespace: "default",
+				Name:      d.configMapName,
+			},
+			Data: d.options.Template,
+		}
+		createdConfigMap, createErr := configMapsClient.Create(context.TODO(), &configMap, metaV1.CreateOptions{})
+		if createErr != nil {
+			return nil, createErr
+		}
+		return createdConfigMap.Data, err
+	}
+	data := d.updateConfigMap(configMapsClient, controllerConfig)
+	return data, nil
+}
+
+func (d *dynamicConfigInternal) updateConfigMap(configMapsClient v1.ConfigMapInterface,
+	stored *corev1.ConfigMap) map[string]string {
+	if d.options.Template == nil {
+		return stored.Data
+	}
+	merged := map[string]string{}
+	for key, value := range stored.Data {
+		if d.options.RemoveUnknownProperties {
+			if _, ok := d.options.Template[key]; ok {
+				merged[key] = value
+			}
+		} else {
+			merged[key] = value
+		}
+	}
+	for key, value := range d.options.Template {
+		if _, ok := stored.Data[key]; !ok {
+			merged[key] = value
+		}
+	}
+
+	stored.Data = merged
+	updated, err := configMapsClient.Update(context.TODO(), stored, metaV1.UpdateOptions{})
+	if err != nil {
+		if d.options.OnErrorCallback != nil {
+			d.options.OnErrorCallback(err, nil)
+		}
+		return stored.Data
+	}
+
+	return updated.Data
+}
+
+func (d *dynamicConfigInternal) startRefreshProcess(clientSet *kubernetes.Clientset) {
+	informerFactory := informers.NewSharedInformerFactoryWithOptions(clientSet, time.Minute)
+	configMapInformer := informerFactory.Core().V1().ConfigMaps().Informer()
+	configMapInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			d.processInformerUpdate(obj)
+		},
+		UpdateFunc: func(oldObj interface{}, obj interface{}) {
+			d.processInformerUpdate(obj)
+		},
+	})
+
+	informerFactory.Start(d.refreshCancel)
+}
+
+func (d *dynamicConfigInternal) processInformerUpdate(obj interface{}) {
+	config := obj.(*corev1.ConfigMap)
+	if config.Name == d.configMapName {
+		_ = d.process(config.Data)
+	}
+}
+
+func (d *dynamicConfigInternal) process(raw map[string]string) error {
+	mapped, err := d.configMapper(raw, d.getCurrent())
+
+	if err != nil {
+		if d.options.OnUpdateCallback != nil {
+			d.options.OnErrorCallback(err, d.current)
+		}
+		return err
+	}
+	current := ConfigState{
+		Raw:       raw,
+		Mapped:    mapped,
+		Timestamp: time.Now(),
```

Comment:
Is there some way to get the "original" timestamp from the config map update, instead of "now" (process time)?

I could imagine `Timestamp` being a source of confusion, because it really represents (currently) "when the config map was processed by this go code" and not "when this config map was updated originally"

Reply:
There is no something like `lastUpdateTimestamp`. There is some field level metadata available, but more complicated to process.